### PR TITLE
librbd: improve journal performance to match expected degredation

### PR DESCRIPTION
**Pull request by dillaman**
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

### CODE REVIEW

**Comment by dillaman** [comment]
Performance results comparing no journal, pre-change, and post-change:

```
	no journal	original	modified
4K	40310.84	7223.73		27427.27
32K	21530.37	3256.39		10284.05
256K	3159.67		564.73		1627.16
4M	223.03		48.03		91.91
```

**Comment by trociny** [comment]
@dillaman could you please rebase?

File: src/journal/ObjectRecorder.cc [hunk 1]
```diff
@@ -349,93 +232,107 @@ void ObjectRecorder::append_overflowed() {
...
+  librados::ObjectWriteOperation op;
+  client::guard_append(&op, m_soft_max_size);
+
+  size_t append_bytes = 0;
+  AppendBuffers append_buffers;
+  for (auto it = m_pending_buffers.begin(); it != m_pending_buffers.end(); ) {
+    auto& future = it->first;
+    auto& bl = it->second;
+    auto size = m_object_bytes + m_in_flight_bytes + append_bytes + bl.length();
+    if (size == m_soft_max_size) {
```
Comment (trociny) [inline src/journal/ObjectRecorder.cc]: @dillaman Just wondering, is there something special in `size == m_soft_max_size` case (in comparison to `>` case) that you added a separate debug message for this. It seems to me we will hardly hit this?

File: src/journal/ObjectRecorder.cc [hunk 2]
```diff
@@ -42,54 +39,36 @@ ObjectRecorder::ObjectRecorder(librados::IoCtx &ioctx, const std::string &oid,
...
-      schedule_append_task();
-    } else {
-      cancel_append_task();
-    }
+    m_pending_buffers.push_back(append_buffer);
+    m_pending_bytes += append_buffer.second.length();
   }
-  return (!m_object_closed && !m_overflowed &&
-          m_size + m_pending_bytes >= m_soft_max_size);
+  m_pending_time = ceph_clock_now();
```
Comment (trociny) [inline src/journal/ObjectRecorder.cc]: Considering an age case, when `flush_bytes`  and `flush_count` limits are set to infinity, and appends are added in time interval less than `flush_age`,  wouldn't it mean that we would never trigger `flush_age`?

File: src/journal/ObjectRecorder.cc [hunk 3]
```diff
@@ -266,18 +162,14 @@ void ObjectRecorder::handle_append_flushed(uint64_t tid, int r) {
...
+    ceph_assert(iter != m_in_flight_appends.end());
+
+    if (r == -EOVERFLOW) {
+      ldout(m_cct, 10) << "append overflowed" << dendl;
+      m_overflowed = true;
 
       // notify of overflow once all in-flight ops are complete
-      if (m_in_flight_tids.empty() && !m_aio_scheduled) {
-        m_append_buffers.splice(m_append_buffers.begin(), m_pending_buffers);
+      if (m_in_flight_tids.empty()) {
```
Comment (trociny) [inline src/journal/ObjectRecorder.cc]: Just thinking, is a scenario possible, when we send two appends (1) and (2); for (1) the osd returns 0, for (2) the osd returns -EOVERFLOW, but (2) completes earlier on the client side? Or there are some guarantee about the order?

File: src/journal/ObjectRecorder.cc [hunk 4]
```diff
@@ -349,93 +232,107 @@ void ObjectRecorder::append_overflowed() {
...
-        m_in_flight_tids.size() >= m_max_in_flight_appends) {
-      ldout(m_cct, 10) << "max in flight appends reached" << dendl;
-      return;
+  if (!force && max_in_flight_appends != 0 &&
+      static_cast<int32_t>(m_in_flight_tids.size()) >= max_in_flight_appends) {
+    ldout(m_cct, 10) << "max in flight appends reached" << dendl;
+    return false;
+  }
+
+  uint64_t append_tid = m_append_tid++;
```
Comment (trociny) [inline src/journal/ObjectRecorder.cc]: nit: why not allocate the new `append_tid` only after the loop below, if `append_bytes > 0`?

File: src/journal/ObjectRecorder.cc (hunk 1 above)
Comment (dillaman) [reply src/journal/ObjectRecorder.cc]: There is a unit test for this case to ensure that it immediately opens a new object set (but still actually appends the record to the current object set)

File: src/journal/ObjectRecorder.cc (hunk 3 above)
Comment (dillaman) [reply src/journal/ObjectRecorder.cc]: Ops are ordered on the wire (both directions) and librados uses a single thread for all callbacks, so order is guaranteed.

File: src/journal/ObjectRecorder.cc (hunk 4 above)
Comment (dillaman) [reply src/journal/ObjectRecorder.cc]: Done

**Pull request merged**